src/Files.App/Filesystem/FilesystemOperations/FileSystemProgress.cs

```diff
@@ -0,0 +1,85 @@
+﻿using Files.App.Helpers;
+using Files.Shared.Enums;
+using System;
+
+namespace Files.App.Filesystem
+{
+    public class FileSystemProgress
+    {
+        private readonly IProgress<FileSystemProgress>? progress;
+        private readonly IntervalSampler sampler;
+        private FileSystemStatusCode? status;
+        private bool criticalReport;
+        private bool enumerationCompleted;
+
+        public FileSystemStatusCode? Status
+        {
+            get => status;
+            set
+            {
+                if (status != value)
+                {
+                    criticalReport = true;
+                }
+                status = value;
+            }
+        }
+        public string? FileName { get; set; }
+        public long TotalSize { get; set; }
+        public long ProcessedSize { get; set; }
+        public long ItemsCount { get; set; }
+        public long ProcessedItemsCount { get; set; }
+        public DateTimeOffset StartTime { get; }
+        public DateTimeOffset CompletedTime { get; private set; }
+        public bool EnumerationCompleted
+        {
+            get => enumerationCompleted;
+            set
+            {
+                if (enumerationCompleted != value)
+                {
+                    criticalReport = true;
+                }
+                enumerationCompleted = value;
+            }
+        }
+        /// <summary>
+        /// Only used when detailed count isn't available.
+        /// </summary>
+        public float? Percentage { get; set; }
```

Comment:
Could it be worth to wrap both the Percentage and the Count inside an object that would adapt its behavior depending on the data avalaible to it?

Reply:
Percentage here effectively overrides the details due to count not available (particularly for `ShellFilesOperation`). I prefer not to wrap them inside an object to keep it simple.